### ISSUE TITLE
Almost never use IDs in CSS

### ISSUE THREAD

**Issue by reubenmoes** [domain7/dev-wiki#3]
Using IDs in CSS almost always ends up with hair loss.  I think we should have a note in here about using classes over IDs whenever we have control over it.  Reasons: 
1. Less hassle with overriding styles.
2. More reusability 

@nathanshubert @eriku @verticalgrain @andmiriam @lancejanocha

**Comment by eriku** [comment]
I'm definitely on board with that. The fact that we now build "modular" websites means that we would only be able to have one module per page... meaning it's not all that modular anymore.

**Comment by nathansh** [comment]
Yeah, I think in general the documentation definitely does favour a classes over ID's approach, but I think we could be more specific in that. I definitely disagree that we should say 'almost never use IDs in CSS'. There definitely is such a trend, but I think it's a bit absolutist. I think the traditional "classes for reusable things, IDs for things that there's only one off" has some value. And you can give an element an ID _and_ a class too. So things like the masthead of the site... I usually do header#masthead for that since there's only one masthead ever. So in HTML it'll have that ID, but I'll also give it a class for any reusable styles. But, if I do want to do something very specific for the masthead, I might have a `#masthead { ... }` in a partial with some very specific, one off css for it. I'm fairly sure this is appropriate use of IDs... all really one-off in nature and clearly such.

Having said that, I wouldn't be opposed to something along the lines of "In general favour classes over IDs. Only use IDs if you have a good, considered reason for doing so, in situations where that styling will never need to be reused. When using IDs, ensure that you aren't creating more specificity than needed for nested declaration blocks." Something like that.

**Comment by reubenmoes** [comment]
I would say in general, relying on styles nested under #header and #footer will often end up in specificity headaches.  Here is a bit of a dumb example:  say you wanted to theme all links based on what section (class added to body) you are in.  Say you did something like assigning all the links in the header a color.  Even though it is listed before the override styles, and the override style is on a more-root DOM element, the link would still be set by the ID class.  I admit that this might actually be a valid use case (to keep the header links a certain color), so this might not be the best example, however I have run into issues like this in the past and it often ends up with !important all over the place or super specific CSS selectors to override things (matching all parent child items).

I would say in general, we should use a CSS Class over an ID, unless there is a strong reason to use an ID.  

```
#header a {
  color: purple;
}
a {
  color: blue;
    .section-blog & {
      color: red;
    }
}

```

**Comment by nathansh** [comment]
I'd say if you _do_ find yourself with a specificity problem, and most definitely if you use !important at all, then you're for sure not using the ID in a useful way. I think what's happening as a trend is instead of focusing on useful vs poor use of IDs people are saying don't use IDs at all, basically as a safety. It's probably a good idea to do that, but I'd just word it more like "favour classes" over "never use IDs because they're evil." Because, it's not really IDs that are the problem so much as how people tend to use them.

**Comment by reubenmoes** [comment]
Maybe if we just add a warning about using IDs and how there are specificity issues with it.

**Comment by nathansh** [comment]
I'd be into that.

**Comment by nathansh** [comment]
Maybe we should also a comment about how you should never _need_ an !important and if you do that's the canary in the coal mine for deeper specificity problems.

**Comment by reubenmoes** [comment]
Maybe we just need a section about specificity where we can list all the specificity issues that you might come across.

**Comment by nathansh** [comment]
Sounds like a good approach

**Comment by eriku** [comment]
Should we also say a rough limit on nesting depth?

**Comment by nathansh** [comment]
I think I have a bit about not nesting unnecessarily deep.

**Comment by nathansh** [comment]
I don't think we can really assign a number to it though. I think there's room for common sense and evaluating it case by case with that one.

**Comment by eriku** [comment]
Basically something like ... If you're nesting 10+ levels then you might need to re-think your work :)

**Comment by nathansh** [comment]
Yeah, I think logical nesting would prevent you from ending up 10 levels deep. We could make that section more strongly worded though.

**Comment by nathansh** [comment]
@eriku http://domain7.github.io/fed-wiki/css/#sass-style-guide_nesting

**Comment by andmiriam** [comment]
Sorry I'm so late to the party on this one, these emails go to an address I don't check as often.
Two thoughts:

1.) I'd be up for saying that you run into specificity issues when you use ID's. I know we saw a bunch of that on Imperial. When we used IDs instead of Classes and then tried to add a color theme on top, we had to jump up and down the element chain to get overrides in place without adding an !important. Specificity is definitely the underlying problem. I'd say "Strongly favour classes over IDs unless you have a very good reason to use one." There are a few situations for one time use elements (and I mean one time use over the whole project) that **should** be IDs. Navigational skip links for example. You really want those to be specific and you don't want anything else to be able to override them.

2.) It seems like nesting might turn into it's own thread here but I'll chime in here too unless we move this elsehwere. Chris Coyier's rule is **"Maximum nesting: 3 levels deep"** and I'm with him. As he points out, deep nesting makes up for crappy selectors that are either _dependent on html structure_ (fragile), _overly specific_ (too powerful) or are just _not reusable_(not useful) . We might want to phrase it "Try not to nest deeper than 3 levels". In fact, blocks longer than 50 lines should be a no-no too. They become harder to understand the _deeper and longer_ they are, and unless there is a really good reason for it, we should avoid it. 

(cc @nathanshubert @eriku @lancejanocha @reubenmoes @verticalgrain )

**Comment by nathansh** [comment]
@andmiriam See my note about the 3 levels, 50 lines rule on Erik's ticket: https://github.com/domain7/fed-wiki/issues/4. tl;dr, I think it's a good guideline but I don't think it's useful as a hard and fast rule. I think good guidelines and good judgment can prevail here.

**Comment by eriku** [comment]
I would say something like **never use ID's for basic styles** and only use them you need to adjust for a specific parent selector.

Example:

``` sass
[role=navigation] {

  li {
    display: inline;
  }

  a {
    color: #fff;

    #override & {
      color: #000;
    }    
  }
}
```

**Comment by nathansh** [comment]
I'm not sure how that example works as negating the use of IDs. Also, I don't think we should style with the attribute selector unless we really need to. Reason being, you want the thing you style to be as immutable as possible. This is why styling classes is better than styling element names; you want to be able to change your markup without breaking your styles. I'd say this might also work for attributes in many cases.

Let's take this statement as a draft: "In general favour classes over IDs. Only use IDs if you have a good, considered reason for doing so, in situations where that styling will never need to be reused. When using IDs, ensure that you aren't creating more specificity than needed for nested declaration blocks." What would this _not_ cover? I think in general people are wanting to be more strict in the language, so maybe we also add something like "In case of doubt, err on the side of obtaining from ID usage."

**Comment by reubenmoes** [comment]
The draft sounds good.  I think it might be good to include my example of the ID not being overridden.  I showed that to RJ yesterday and he had no idea about that, so I see the benefit of showing the case of how using an ID is is similar to using !important. 

```
#header a {
  color: purple;
}
a {
  color: blue;
    .section-blog & {
      color: red;
    }
}
```

**Comment by eriku** [comment]
**Draft 2**
Favour classes over IDs. ID selectors are more specific, they are allowed but use them sparingly and appropriately. Remember that IDs are harder to override, almost the same as using `!important`.

Example:

``` scss
#header a {
  // This will not be overridden even though this declaration comes first
  color: purple;
}


a {
  color: blue;

    .section-blog & {
      color: red;
    }
}
```

**Comment by nathansh** [comment]
Looks good, this avoids the dogma I was originally concerned about.

**Comment by reubenmoes** [comment]
looks good to me!

**Comment by nathansh** [comment]
Though maybe we should just considers this into a section on specificity. The problem in this example could exist without an ID too if the previous selector had higher specificity.

**Comment by nathansh** [comment]
Er, consolidate.

**Comment by eriku** [comment]
Good point... Should we just put a comment in the Nesting Depth section?

**Comment by nathansh** [comment]
Makes sense. Nesting is practically speaking where you're going to loose control of specificity.

**Comment by reubenmoes** [comment]
@nathanshubert  So are we not adding this to the wiki? 
https://github.com/domain7/dev-wiki/issues/3#issuecomment-66796388

**Comment by nathansh** [comment]
Just figured there's been no activity on this for 4 months, but I think we're all on the same page if someone wants to do a pull request.

**Comment by reubenmoes** [comment]
I'll do it soon

**Comment by nathansh** [comment]
If the section's about specificity, it'd be important to mention this pattern as it really helps with specificity while organizing code in to a hierarchical order:

```
.module_name {
   &-item { ... }
   &-other_item { .. could be nested in .module_name-item, but you don't need the specificity .. }
}
```
